File: examples/benchmarks/benchmarks.lua
local socket = require('socket');
local posix = require('posix');

local Timer = {};
Timer.__index = Timer;

function Timer.new()
  local timer = {
    timestamp = os.time(),
    elapsed = 0,
  };

  setmetatable(timer, Timer);

  return timer;
end

function Timer:start()
  self.timestamp = socket.gettime();
end

function Timer:stop()
  if self.timestamp ~= nil then
    self.elapsed = self.elapsed + (socket.gettime() - self.timestamp)
    self.timestamp = nil
  end
end

function benchmark(descriptor, reps, b)
  if string.match(descriptor, "DISABLED") then
    return;
  end
  print(string.format("running %s", descriptor));
  local timer = Timer.new();

  for i = 0, reps do
    b(timer, i);
  end

  local elapsed = timer.elapsed;

  print(string.format("%s took %f seconds, or %f per exec", descriptor, elapsed, elapsed / reps))
end

local REPS = 10000000;

X = 0;

benchmark('simple math', REPS, function(timer, i)
  timer:start();
  i = i + 1;
  X = X + i + i * i / i % i;
  timer:stop();
end);

function simple_function()
end

benchmark('function calls', REPS, function(timer, _)
  timer:start();
  simple_function();
  timer:stop();
end);

OBJ = { foobarbaz = nil };

benchmark('global & member access', REPS, function(timer, _)
  timer:start();
  local _ = OBJ.foobarbaz;
  timer:stop();
end)

File: examples/benchmarks/benchmarks.py
import time

class Timer:
  def __init__(self) -> None:
    self.elapsed = 0
    self.timestamp = time.time()

  def start(self):
    self.timestamp = time.time()

  def stop(self):
    if self.timestamp != None:
      self.elapsed += (time.time() - self.timestamp)
      self.timestamp = None

def benchmark(descriptor: str, reps: int, f):
  if not descriptor.startswith("DISABLED"):
    print(f"running {descriptor}")

    timer = Timer()

    for i in range(0, reps):
      f(timer, i)

    elapsed = timer.elapsed

    print(f"{descriptor} took {elapsed} seconds, or {elapsed / reps} per exec")

REPS = 10_000_000

x = 0

def simple_math(timer: Timer, i: int):
  timer.start()
  global x
  i += 1
  x += i + i * i / i % i
  timer.stop()

benchmark('simple math', REPS, simple_math)

def simple_function():
  pass

def function_calls(timer: Timer, i: int):
  timer.start()
  simple_function()
  timer.stop()

benchmark('function calls', REPS, function_calls)

class FakeObj:
  def __init__(self) -> None:
    self.foobarbaz = None

OBJ = FakeObj()

def global_member_access(timer: Timer, i: int):
  timer.start()
  global OBJ
  OBJ.foobarbaz
  timer.stop()

benchmark('global & member access', REPS, global_member_access)
